File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        n = len(nums)
        results = [1] * n

        prefix = 1
        for i in range(n):
            results[i] = prefix
            prefix = prefix * nums[i]
        
        postfix = 1
        for j in range(n - 1, -1, -1):
            results[j] *= postfix
            postfix *= nums[j]
        
        return results

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) 
    {
        int n = nums.length;
        int [] result = new int[n];
        int curr = 1;

        Arrays.fill(result, 1);

        for(int i = 0; i < n; i++) {
            result[i] *= curr;
            curr *= nums[i];
        }
        curr = 1;
        for(int i = n - 1; i >= 0; i--) {
            result[i] *= curr;
            curr *= nums[i];
        }

        return result;
    }
}